lib/Dialect/HW/HWOps.cpp

```diff
@@ -747,8 +793,14 @@ void HWModuleOp::build(OpBuilder &builder, OperationState &result,
   bodyRegion->push_back(body);
 
   // Add arguments to the body block.
-  for (auto elt : ports.inputs)
-    body->addArgument(elt.type, builder.getUnknownLoc());
+  auto unknownLoc = builder.getUnknownLoc();
+  for (auto port : ports.inputs) {
+    auto loc = port.loc ? Location(port.loc) : unknownLoc;
+    auto type = port.type;
+    if (port.isInOut() && !type.isa<InOutType>())
+      type = InOutType::get(type);
+    body->addArgument(type, loc);
+  }
```

Comment:
Could you explain why this change was necessary?

Reply:
The first is more obvious, the block argument should be using the port location specified in the PortInfo.  In the HW dialect the LocationAttr in the PortInfo structure is nullable (does not default to UnknownLoc), so I handle that case here. In the FIRRTL dialect's PortInfo, the Location is not nullable, which I think is better, but it would make this change a lot more intrusive to do the same in the HW dialect.

The second thing is that I added a verifier that ensured the block argument locations of a module match the ArgLocs attribute on the module.  While I was there I added a verifier that the function type of the module matched the block argument types, and discovered that this verifier actually fails. It turns out when we create the PortInfo object if the port is a HW::InOutType,[ we strip it and set the isInOut field to true](https://github.com/llvm/circt/blob/main/lib/Dialect/HW/HWOps.cpp#L886).  When we read in the PortInfo object [we respect this field while constructing the function type](https://github.com/llvm/circt/blob/main/lib/Dialect/HW/HWOps.cpp#L559), but were not respecting the `isInOut` [when creating the block arguments](https://github.com/llvm/circt/blob/main/lib/Dialect/HW/HWOps.cpp#L751). This is a pretty bad bug and could be moved out into its own change, I just haven't mustered the energy to do it yet.

Reply:
Moved out into this change: https://github.com/llvm/circt/pull/4583